### PR TITLE
[#5020] Implemented new resource administration API (master)

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: lib/core/include/resource_administration.hpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        auto id() const noexcept -> const std::string&             { return id_; }
+        auto name() const noexcept -> const std::string&           { return name_; }
+        auto type() const noexcept -> const std::string&           { return type_; }
+        auto zone_name() const noexcept -> const std::string&      { return zone_name_; }
+        auto host_name() const noexcept -> const std::string&      { return host_name_; }
+        auto vault_path() const noexcept -> const std::string&     { return vault_path_; }
+        auto status() const noexcept -> resource_status            { return status_; }
+        auto context_string() const noexcept -> const std::string& { return context_string_; }
+        auto comments() const noexcept -> const std::string&       { return comments_; }
+        auto parent_id() const noexcept -> const std::string&      { return parent_id_; }
```
Comment (trel) [inline lib/core/include/resource_administration.hpp:81]: please consider a handy lookup for parent_name() as well - could be best elsewhere.

File: lib/core/src/resource_administration.cpp [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+#include "rodsErrorTable.h"
+
+#include "fmt/format.h"
+
+namespace irods::experimental::administration::NAMESPACE_IMPL
+{
+    auto add_resource(RxComm& _comm, const registration_info& _info) -> std::error_code
+    {
+        const auto vault_path = fmt::format("{}:{}", _info.host_name, _info.vault_path);
```
Comment (trel) [inline lib/core/src/resource_administration.cpp:28]: might need a slightly different name... 'location' within this scope could be good enough

File: lib/core/src/resource_administration.cpp [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    auto info(RxComm& _comm, const resource_name_type _name) -> std::tuple<std::error_code, std::optional<resource_info>>
+    {
+        const auto to_resource_status = [](const std::string& _status)
+        {
+            // clang-format off
+            if (_status == "up")   return resource_status::up;
+            if (_status == "down") return resource_status::down;
+            if (_status == "auto") return resource_status::auto_detect;
+            // clang-format on
```
Comment (trel) [inline lib/core/src/resource_administration.cpp:111]: i know we're trying to get things consistent and easy to hold.   pretty sure the logic a bit deeper is just looking for the substring 'down', so will need to make sure that's considered here.

File: lib/core/src/resource_administration.cpp [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    auto rebalance(RxComm& _comm, const resource_name_type _name) -> std::error_code
+    {
+        generalAdminInp_t input{};
+        input.arg0 = "modify";
+        input.arg1 = "resource";
+        input.arg2 = _name.data();  // resource name
+        input.arg3 = "rebalance";   // option
+        input.arg4 = "";            // new value
```
Comment (trel) [inline lib/core/src/resource_administration.cpp:257]: arg4 might need to be... nothing?

File: lib/core/include/resource_administration.hpp (hunk 1 above)
Comment (korydraughn) [reply lib/core/include/resource_administration.hpp:82]: done.

File: lib/core/src/resource_administration.cpp (hunk 2 above)
Comment (korydraughn) [reply lib/core/src/resource_administration.cpp]: done.

File: lib/core/src/resource_administration.cpp (hunk 4 above)
Comment (korydraughn) [reply lib/core/src/resource_administration.cpp:262]: this has been resolved.

File: lib/core/src/resource_administration.cpp (hunk 3 above)
Comment (korydraughn) [reply lib/core/src/resource_administration.cpp:116]: Following our discussion, I will remove `auto` from the available status values.

File: CMakeLists.txt [hunk 5]
```diff
@@ -355,6 +355,44 @@ set_target_properties(irods_user_administration_server PROPERTIES CXX_STANDARD $
...
+target_include_directories(
+  irods_resource_administration_server
+  PRIVATE
+  ${CMAKE_BINARY_DIR}/lib/core/include
+  ${CMAKE_SOURCE_DIR}/lib/core/include
+  ${CMAKE_SOURCE_DIR}/lib/api/include
+  ${CMAKE_SOURCE_DIR}/server/core/include
+  ${CMAKE_SOURCE_DIR}/server/api/include
+  ${IRODS_EXTERNALS_FULLPATH_BOOST}/include
+  ${IRODS_EXTERNALS_FULLPATH_FMT}/include
```
Comment (korydraughn) [inline CMakeLists.txt:390]: verify that all of these are needed.

**Comment by korydraughn** [comment]
Added documentation and #

File: lib/core/include/resource_administration.hpp [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+    /// A type that holds information about an existing resource.
+    class resource_info
+    {
+    public:
+        // clang-format off
+        auto id() const noexcept -> const std::string&                       { return id_; }
+        auto name() const noexcept -> const std::string&                     { return name_; }
+        auto type() const noexcept -> const std::string&                     { return type_; }
+        auto zone_name() const noexcept -> const std::string&                { return zone_name_; }
```
[the remark this answers is not in the document]
Comment (trel) [reply lib/core/include/resource_administration.hpp:82]: we'll investigate later... merging...

**Pull request merged**